b2/src/main/java/org/jclouds/b2/filters/RequestAuthorization.java

```diff
@@ -0,0 +1,91 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.labs.b2.filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.jclouds.Constants;
+import org.jclouds.domain.Credentials;
+import org.jclouds.http.HttpException;
+import org.jclouds.http.HttpRequest;
+import org.jclouds.http.HttpRequestFilter;
+import org.jclouds.http.HttpUtils;
+import org.jclouds.labs.b2.B2Client;
+import org.jclouds.labs.b2.domain.Authorization;
+import org.jclouds.location.Provider;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Function;
+import com.google.common.base.Supplier;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+@Singleton
+public class RequestAuthorization implements HttpRequestFilter {
+   @Resource
+   @Named(Constants.LOGGER_SIGNATURE)
+   Logger signatureLog = Logger.NULL;
+
+   private final Supplier<Credentials> creds;
+   private final HttpUtils utils;
+   private final LoadingCache<Credentials, Authorization> cache;
+
+   @Inject
+   public RequestAuthorization(@Provider Supplier<Credentials> creds, final B2Client b2Client, HttpUtils utils) {
+      this.creds = checkNotNull(creds, "creds");
+      this.utils = checkNotNull(utils, "utils");
+      this.cache = CacheBuilder.newBuilder()
+            .expireAfterWrite(23, TimeUnit.HOURS)
+            .build(CacheLoader.from(new Function<Credentials, Authorization>() {
+               @Override
+               public Authorization apply(Credentials creds) {
+                  return b2Client.getAuthorizationApi().authorizeAccount(creds.identity, creds.credential);
+               }
+            }));
+   }
+
+   @Override
+   public HttpRequest filter(HttpRequest request) throws HttpException {
+      utils.logRequest(signatureLog, request, ">>");
+      Credentials creds = this.creds.get();
+      Authorization auth = cache.getUnchecked(creds);
+
+      // Replace fixed URL with API URL, except for unit tests tests
+      URI endpoint = request.getEndpoint();
+      if (!endpoint.getHost().equals("localhost")) {
+         endpoint = URI.create(auth.apiUrl() + endpoint.getPath() + (endpoint.getQuery() == null ? "" : "?" + endpoint.getQuery()));
+      }
+
+      request = request.toBuilder()
+            .endpoint(endpoint)
+            // TODO: squirrel away accountId in bogus header so ReplacePayloadAccountId can use it
```

Comment:
B2 requires the accountId in the JSON payload of requests but we don't want to include it in API calls.  Can I more cleanly pass this information to `ReplacePayloadAccountId`?